File: apps/lab_admin_web/src/services/nominatimGeocode.ts
/** Forward geocoding via OSM Nominatim (no API key). Respect their usage policy in production. */

export type GeocodeHit = { lat: number; lng: number }

type NominatimRow = { lat?: string; lon?: string }

export async function nominatimSearch(query: string, signal?: AbortSignal): Promise<GeocodeHit | null> {
  const q = query.trim()
  if (q.length < 4) return null

  const url = new URL('https://nominatim.openstreetmap.org/search')
  url.searchParams.set('format', 'json')
  url.searchParams.set('limit', '1')
  url.searchParams.set('q', q)

  const res = await fetch(url.toString(), {
    signal,
    headers: {
      Accept: 'application/json',
      'Accept-Language': 'en',
      'User-Agent': 'MedLabSmartAdmin/1.0 (lab_admin_web)',
    },
  })

  if (!res.ok) return null

  const data = (await res.json()) as NominatimRow[]
  if (!Array.isArray(data) || data.length === 0) return null

  const lat = Number.parseFloat(String(data[0].lat ?? ''))
  const lng = Number.parseFloat(String(data[0].lon ?? ''))
  if (!Number.isFinite(lat) || !Number.isFinite(lng)) return null

  return { lat, lng }
}

/** Reverse geocoding: coordinates → single-line address (`display_name`). */
export async function nominatimReverse(
  lat: number,
  lng: number,
  signal?: AbortSignal,
): Promise<string | null> {
  if (!Number.isFinite(lat) || !Number.isFinite(lng)) return null

  const url = new URL('https://nominatim.openstreetmap.org/reverse')
  url.searchParams.set('format', 'json')
  url.searchParams.set('lat', String(lat))
  url.searchParams.set('lon', String(lng))

  const res = await fetch(url.toString(), {
    signal,
    headers: {
      Accept: 'application/json',
      'Accept-Language': 'en',
      'User-Agent': 'MedLabSmartAdmin/1.0 (lab_admin_web)',
    },
  })

  if (!res.ok) return null

  const data = (await res.json()) as { display_name?: string }
  const line = typeof data.display_name === 'string' ? data.display_name.trim() : ''
  return line || null
}

File: apps/lab_mobile_app/lib/services/nominatim_geocode.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

/// OSM Nominatim (no API key). Use sparingly; include a stable User-Agent per their policy.
class NominatimGeocode {
  NominatimGeocode._();

  static const _userAgent = 'MedLabSmartMobile/1.0 (lab_patient_app)';

  static Future<String?> reverse(double lat, double lng) async {
    if (!lat.isFinite || !lng.isFinite) return null;
    final uri = Uri.parse('https://nominatim.openstreetmap.org/reverse').replace(
      queryParameters: <String, String>{
        'format': 'json',
        'lat': '$lat',
        'lon': '$lng',
      },
    );
    final res = await http.get(
      uri,
      headers: const {
        'Accept': 'application/json',
        'Accept-Language': 'en',
        'User-Agent': _userAgent,
      },
    );
    if (res.statusCode < 200 || res.statusCode >= 300) return null;
    try {
      final data = jsonDecode(res.body);
      if (data is! Map<String, dynamic>) return null;
      final line = data['display_name'];
      if (line is! String) return null;
      final t = line.trim();
      return t.isEmpty ? null : t;
    } catch (_) {
      return null;
    }
  }

  /// Forward search: free-text query → first hit coordinates (optional helper).
  static Future<({double lat, double lng, String? displayName})?> search(String query) async {
    final q = query.trim();
    if (q.length < 4) return null;
    final uri = Uri.parse('https://nominatim.openstreetmap.org/search').replace(
      queryParameters: <String, String>{
        'format': 'json',
        'limit': '1',
        'q': q,
      },
    );
    final res = await http.get(
      uri,
      headers: const {
        'Accept': 'application/json',
        'Accept-Language': 'en',
        'User-Agent': _userAgent,
      },
    );
    if (res.statusCode < 200 || res.statusCode >= 300) return null;
    try {
      final data = jsonDecode(res.body);
      if (data is! List || data.isEmpty) return null;
      final row = data.first;
      if (row is! Map<String, dynamic>) return null;
      final lat = double.tryParse('${row['lat'] ?? ''}');
      final lon = double.tryParse('${row['lon'] ?? ''}');
      if (lat == null || lon == null) return null;
      if (!lat.isFinite || !lon.isFinite) return null;
      final dn = row['display_name'];
      final displayName = dn is String && dn.trim().isNotEmpty ? dn.trim() : null;
      return (lat: lat, lng: lon, displayName: displayName);
    } catch (_) {
      return null;
    }
  }
}
